File: core/src/test/java/com/pipai/adv/backend/battle/engine/NormalAttackTest.java
package com.pipai.adv.backend.battle.engine;

import com.pipai.adv.backend.battle.domain.*;
import com.pipai.adv.npc.Npc;
import com.pipai.adv.npc.NpcList;
import com.pipai.adv.save.AdvSave;
import com.pipai.test.fixtures.TestFixturesKt;
import org.junit.Assert;
import org.junit.Test;

import java.util.List;

public class NormalAttackTest {

    @Test
    public void testMeleeNormalAttack() {
        AdvSave save = new AdvSave();
        NpcList npcList = new NpcList();
        BattleMap map = BattleMap.Factory.createBattleMap(4, 4);
        InventoryItem.WeaponInstance sword = TestFixturesKt.swordFixture();
        Npc attacker = TestFixturesKt.npcFromStats(new UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3),
                sword);
        int attackerId = npcList.addNpc(attacker);
        Npc target = TestFixturesKt.npcFromStats(new UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3),
                null);
        int targetId = npcList.addNpc(target);
        map.getCell(0, 0).setFullEnvObject(new FullEnvObject.NpcEnvObject(attackerId, Team.PLAYER, EnvObjTilesetMetadata.NONE));
        map.getCell(1, 1).setFullEnvObject(new FullEnvObject.NpcEnvObject(targetId, Team.PLAYER, EnvObjTilesetMetadata.NONE));

        BattleBackend backend = new BattleBackend(save, npcList, map);

        NormalAttackCommand cmd = new NormalAttackCommand(attackerId, targetId, sword);

        Assert.assertTrue(backend.canBeExecuted(cmd).getExecutable());

        List<BattleLogEvent> events = backend.execute(cmd);
        Assert.assertTrue(events.stream().anyMatch(it -> it instanceof DamageEvent));
        DamageEvent damageEvent = (DamageEvent) events.stream().filter(it -> it instanceof DamageEvent).findFirst().get();
        Assert.assertEquals(targetId, damageEvent.getNpcId());
        Assert.assertEquals(100 - damageEvent.getDamage(), target.getUnitInstance().getHp());
    }

    @Test
    public void testMeleeNormalAttackOutOfRange() {
        AdvSave save = new AdvSave();
        NpcList npcList = new NpcList();
        BattleMap map = BattleMap.Factory.createBattleMap(4, 4);
        InventoryItem.WeaponInstance sword = TestFixturesKt.swordFixture();
        Npc attacker = TestFixturesKt.npcFromStats(new UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3),
                sword);
        int attackerId = npcList.addNpc(attacker);
        Npc target = TestFixturesKt.npcFromStats(new UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3),
                null);
        int targetId = npcList.addNpc(target);
        map.getCell(0, 0).setFullEnvObject(new FullEnvObject.NpcEnvObject(attackerId, Team.PLAYER, EnvObjTilesetMetadata.NONE));
        map.getCell(2, 0).setFullEnvObject(new FullEnvObject.NpcEnvObject(targetId, Team.PLAYER, EnvObjTilesetMetadata.NONE));

        BattleBackend backend = new BattleBackend(save, npcList, map);

        NormalAttackCommand cmd = new NormalAttackCommand(attackerId, targetId, sword);

        ExecutableStatus status = backend.canBeExecuted(cmd);
        Assert.assertFalse(status.getExecutable());
        Assert.assertEquals("Attacking distance is too great", status.getReason());
    }

    @Test
    public void testBowRangedNormalAttack() {
        AdvSave save = new AdvSave();
        NpcList npcList = new NpcList();
        BattleMap map = BattleMap.Factory.createBattleMap(4, 4);
        InventoryItem.WeaponInstance bow = TestFixturesKt.bowFixture();
        Npc attacker = TestFixturesKt.npcFromStats(new UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3),
                bow);
        int attackerId = npcList.addNpc(attacker);
        Npc target = TestFixturesKt.npcFromStats(new UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3),
                null);
        int targetId = npcList.addNpc(target);
        map.getCell(0, 0).setFullEnvObject(new FullEnvObject.NpcEnvObject(attackerId, Team.PLAYER, EnvObjTilesetMetadata.NONE));
        map.getCell(2, 0).setFullEnvObject(new FullEnvObject.NpcEnvObject(targetId, Team.PLAYER, EnvObjTilesetMetadata.NONE));

        BattleBackend backend = new BattleBackend(save, npcList, map);

        NormalAttackCommand cmd = new NormalAttackCommand(attackerId, targetId, bow);

        Assert.assertTrue(backend.canBeExecuted(cmd).getExecutable());

        List<BattleLogEvent> events = backend.execute(cmd);
        Assert.assertTrue(events.stream().anyMatch(it -> it instanceof NormalAttackEvent));
        NormalAttackEvent normalAttackEvent = (NormalAttackEvent) events.stream().filter(it -> it instanceof NormalAttackEvent).findFirst().get();
        Assert.assertEquals(attackerId, normalAttackEvent.getAttackerId());
        Assert.assertEquals(targetId, normalAttackEvent.getTargetId());
        Assert.assertEquals(bow, normalAttackEvent.getWeapon());

        Assert.assertTrue(events.stream().anyMatch(it -> it instanceof DamageEvent));
        DamageEvent damageEvent = (DamageEvent) events.stream().filter(it -> it instanceof DamageEvent).findFirst().get();
        Assert.assertEquals(targetId, damageEvent.getNpcId());
        Assert.assertEquals(100 - damageEvent.getDamage(), target.getUnitInstance().getHp());

        Assert.assertTrue(events.stream().anyMatch(it -> it instanceof AmmoChangeEvent));
        AmmoChangeEvent ammoChangeEvent = (AmmoChangeEvent) events.stream().filter(it -> it instanceof AmmoChangeEvent).findFirst().get();
        Assert.assertEquals(2, ammoChangeEvent.getNewAmount());
        Assert.assertEquals(attackerId, ammoChangeEvent.getNpcId());
        Assert.assertEquals(2, bow.getAmmo());
    }

    @Test
    public void testBowRangedNormalAttackOutOfRange() {
        AdvSave save = new AdvSave();
        NpcList npcList = new NpcList();
        BattleMap map = BattleMap.Factory.createBattleMap((int) BattleBackend.RANGED_WEAPON_DISTANCE + 1, 1);
        InventoryItem.WeaponInstance sword = TestFixturesKt.swordFixture();
        Npc attacker = TestFixturesKt.npcFromStats(new UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3),
                sword);
        int attackerId = npcList.addNpc(attacker);
        Npc target = TestFixturesKt.npcFromStats(new UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3),
                null);
        int targetId = npcList.addNpc(target);
        map.getCell(0, 0).setFullEnvObject(new FullEnvObject.NpcEnvObject(attackerId, Team.PLAYER, EnvObjTilesetMetadata.NONE));
        map.getCell((int) BattleBackend.RANGED_WEAPON_DISTANCE, 0).setFullEnvObject(new FullEnvObject.NpcEnvObject(targetId, Team.PLAYER, EnvObjTilesetMetadata.NONE));

        BattleBackend backend = new BattleBackend(save, npcList, map);

        NormalAttackCommand cmd = new NormalAttackCommand(attackerId, targetId, sword);

        ExecutableStatus status = backend.canBeExecuted(cmd);
        Assert.assertFalse(status.getExecutable());
        Assert.assertEquals("Attacking distance is too great", status.getReason());
    }

    @Test
    public void testBowOutOfAmmoNormalAttack() {
        AdvSave save = new AdvSave();
        NpcList npcList = new NpcList();
        BattleMap map = BattleMap.Factory.createBattleMap(4, 4);
        InventoryItem.WeaponInstance bow = TestFixturesKt.bowFixture();
        bow.setAmmo(0);
        Npc attacker = TestFixturesKt.npcFromStats(new UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3),
                bow);
        int attackerId = npcList.addNpc(attacker);
        Npc target = TestFixturesKt.npcFromStats(new UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3),
                null);
        int targetId = npcList.addNpc(target);
        map.getCell(0, 0).setFullEnvObject(new FullEnvObject.NpcEnvObject(attackerId, Team.PLAYER, EnvObjTilesetMetadata.NONE));
        map.getCell(2, 0).setFullEnvObject(new FullEnvObject.NpcEnvObject(targetId, Team.PLAYER, EnvObjTilesetMetadata.NONE));

        BattleBackend backend = new BattleBackend(save, npcList, map);

        NormalAttackCommand cmd = new NormalAttackCommand(attackerId, targetId, bow);

        ExecutableStatus status = backend.canBeExecuted(cmd);
        Assert.assertFalse(status.getExecutable());
        Assert.assertEquals("This weapon needs to be reloaded", status.getReason());
    }
}

File: core/src/test/java/com/pipai/adv/backend/battle/engine/NormalAttackTest.kt
package com.pipai.adv.backend.battle.engine

import com.pipai.adv.backend.battle.domain.*
import com.pipai.adv.backend.battle.engine.commands.NormalAttackCommand
import com.pipai.adv.backend.battle.engine.log.DamageEvent
import com.pipai.adv.backend.battle.engine.log.NormalAttackEvent
import com.pipai.adv.domain.Npc
import com.pipai.adv.utils.AutoIncrementIdMap
import com.pipai.test.fixtures.bowFixture
import com.pipai.test.fixtures.npcFromStats
import com.pipai.test.fixtures.swordFixture
import com.pipai.test.libgdx.GdxMockedTest
import com.pipai.test.libgdx.generateBackend
import org.junit.Assert
import org.junit.Test

class NormalAttackTest : GdxMockedTest() {

    @Test
    fun testMeleeNormalAttack() {
        val npcList = AutoIncrementIdMap<Npc>()
        val envObjList = AutoIncrementIdMap<EnvObject>()

        val map = BattleMap.createBattleMap(4, 4)
        val sword = swordFixture()
        val attacker = npcFromStats(UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3),
                sword)
        val attackerId = npcList.add(attacker)
        val target = npcFromStats(UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3),
                null)
        val targetId = npcList.add(target)
        map.getCell(0, 0).fullEnvObjId = envObjList.add(NpcEnvObject(attackerId, Team.PLAYER, EnvObjTilesetMetadata.NONE))
        map.getCell(1, 1).fullEnvObjId = envObjList.add(NpcEnvObject(targetId, Team.PLAYER, EnvObjTilesetMetadata.NONE))

        val backend = generateBackend(npcList, envObjList, map)

        val cmd = NormalAttackCommand(attackerId, targetId)

        Assert.assertTrue(backend.canBeExecuted(cmd).executable)

        val events = backend.execute(cmd)
        Assert.assertTrue(events.stream().anyMatch { it -> it is DamageEvent })
        val (npcId, _, damage) = events.stream().filter { it -> it is DamageEvent }.findFirst().get() as DamageEvent
        Assert.assertEquals(targetId.toLong(), npcId.toLong())
        Assert.assertEquals((100 - damage).toLong(), target.unitInstance.hp.toLong())

        Assert.assertEquals(0, backend.getNpcAp(attackerId))
    }

    @Test
    fun testMeleeNormalAttackOutOfRange() {
        val npcList = AutoIncrementIdMap<Npc>()
        val envObjList = AutoIncrementIdMap<EnvObject>()

        val map = BattleMap.createBattleMap(4, 4)
        val sword = swordFixture()
        val attacker = npcFromStats(UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3),
                sword)
        val attackerId = npcList.add(attacker)
        val target = npcFromStats(UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3), null)
        val targetId = npcList.add(target)
        map.getCell(0, 0).fullEnvObjId = envObjList.add(NpcEnvObject(attackerId, Team.PLAYER, EnvObjTilesetMetadata.NONE))
        map.getCell(2, 0).fullEnvObjId = envObjList.add(NpcEnvObject(targetId, Team.PLAYER, EnvObjTilesetMetadata.NONE))

        val backend = generateBackend(npcList, envObjList, map)

        val cmd = NormalAttackCommand(attackerId, targetId)

        val (executable, reason) = backend.canBeExecuted(cmd)
        Assert.assertFalse(executable)
        Assert.assertEquals("Attacking distance is too great", reason)
    }

    @Test
    fun testBowRangedNormalAttack() {
        val npcList = AutoIncrementIdMap<Npc>()
        val envObjList = AutoIncrementIdMap<EnvObject>()

        val map = BattleMap.createBattleMap(4, 4)
        val bow = bowFixture()
        val attacker = npcFromStats(UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3),
                bow)
        val attackerId = npcList.add(attacker)
        val target = npcFromStats(UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3), null)
        val targetId = npcList.add(target)
        map.getCell(0, 0).fullEnvObjId = envObjList.add(NpcEnvObject(attackerId, Team.PLAYER, EnvObjTilesetMetadata.NONE))
        map.getCell(2, 0).fullEnvObjId = envObjList.add(NpcEnvObject(targetId, Team.PLAYER, EnvObjTilesetMetadata.NONE))

        val backend = generateBackend(npcList, envObjList, map)

        val cmd = NormalAttackCommand(attackerId, targetId)

        Assert.assertTrue(backend.canBeExecuted(cmd).executable)

        val events = backend.execute(cmd)
        Assert.assertTrue(events.stream().anyMatch { it -> it is NormalAttackEvent })
        val (attackerId1, _, targetId1, _, weapon) = events.stream().filter { it -> it is NormalAttackEvent }.findFirst().get() as NormalAttackEvent
        Assert.assertEquals(attackerId.toLong(), attackerId1.toLong())
        Assert.assertEquals(targetId.toLong(), targetId1.toLong())
        Assert.assertEquals(bow, weapon)

        Assert.assertTrue(events.stream().anyMatch { it -> it is DamageEvent })
        val (npcId, _, damage) = events.stream().filter { it -> it is DamageEvent }.findFirst().get() as DamageEvent
        Assert.assertEquals(targetId.toLong(), npcId.toLong())
        Assert.assertEquals((100 - damage).toLong(), target.unitInstance.hp.toLong())

        Assert.assertEquals(0, backend.getNpcAp(attackerId))
    }

    @Test
    fun testBowRangedNormalAttackOutOfRange() {
        val npcList = AutoIncrementIdMap<Npc>()
        val envObjList = AutoIncrementIdMap<EnvObject>()

        val map = BattleMap.createBattleMap(BattleBackend.RANGED_WEAPON_DISTANCE.toInt() + 1, 1)
        val sword = swordFixture()
        val attacker = npcFromStats(UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3),
                sword)
        val attackerId = npcList.add(attacker)
        val target = npcFromStats(UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3), null)
        val targetId = npcList.add(target)
        map.getCell(0, 0).fullEnvObjId = envObjList.add(NpcEnvObject(attackerId, Team.PLAYER, EnvObjTilesetMetadata.NONE))
        map.getCell(BattleBackend.RANGED_WEAPON_DISTANCE.toInt(), 0)
                .fullEnvObjId = envObjList.add(NpcEnvObject(targetId, Team.PLAYER, EnvObjTilesetMetadata.NONE))

        val backend = generateBackend(npcList, envObjList, map)

        val cmd = NormalAttackCommand(attackerId, targetId)

        val (executable, reason) = backend.canBeExecuted(cmd)
        Assert.assertFalse(executable)
        Assert.assertEquals("Attacking distance is too great", reason)
    }
}
